python/tank/bootstrap/async.py

```diff
@@ -0,0 +1,278 @@
+        """
+        Starts the asynchronous bootstrap logic.
+        """
+
+        # Start the QThread object event loop in its new thread context.
+        self._thread.start()
+
+    @QtCore.Slot(float, str, int, int)
+    def _progress_bootstrap(self, progress_value, message, current_index, maximum_index):
+        """
+        Callback slot that reports back on the toolkit and engine bootstrap progress.
+
+        :param progress_value: Current progress value, ranging from 0.0 to 1.0.
+        :param message: Progress message to report.
+        :param current_index: Optional current item number being looped over, or None.
+        :param maximum_index: Optional maximum item number being looped over, or None.
+        """
+
+        if self._progress_callback:
+            self._progress_callback(progress_value, message, current_index, maximum_index)
+
+    @QtCore.Slot(Sgtk)
+    def _complete_bootstrap(self, toolkit):
+        """
+        Callback slot that handles cleanup after successful completion of the toolkit bootstrap.
+
+        :param toolkit: Bootstrapped :class:`~sgtk.Sgtk` instance.
+        """
+
+        # Make the QThread object exit its event loop.
+        self._thread.quit()
+
+        # Make the worker operate with the main thread affinity
+        # where the main event loop can handle its deletion.
+        self._worker.moveToThread(self.thread())
+
+        try:
+
+            # Ladies and Gentlemen, start your engines!
+            engine = self._toolkit_manager._start_engine(toolkit, self._engine_name, self._entity)
+
+        except Exception, exception:
+
+            if self._failed_callback:
+                # Handle cleanup after failed completion of the engine startup.
+                self._failed_callback(self._toolkit_manager.ENGINE_STARTUP_PHASE, exception)
+
+            return
+
+        if self._completed_callback:
+            # Handle cleanup after successful completion of the engine startup.
+            self._completed_callback(engine)
+
+    @QtCore.Slot(Exception)
+    def _fail_bootstrap(self, exception):
+        """
+        Callback slot that handles cleanup after failed completion of the toolkit bootstrap.
+
+        :param exception: Exception raised while bootstrapping the toolkit.
+        """
+
+        # Make the QThread object exit its event loop.
+        self._thread.quit()
+
+        # Make the worker operate with the main thread affinity
+        # where the main event loop can handle its deletion.
+        self._worker.moveToThread(self.thread())
+
+        if self._failed_callback:
+            # Handle cleanup after failed completion of the toolkit bootstrap.
+            self._failed_callback(self._toolkit_manager.TOOLKIT_BOOTSTRAP_PHASE, exception)
+
+
+class _BootstrapToolkitWorker(QtCore.QObject):
+    """
+    Bootstrap worker that can bootstrap an :class:`~sgtk.Sgtk` instance asynchronously in a background thread.
+
+    This class defines 3 Qt signals:
```

Comment:
minor nitpick: we have a slightly different syntax we normally use for signal docstrings (see wiki). Example: https://github.com/shotgunsoftware/tk-framework-qtwidgets/blob/9d35259cadcdc7d9a53e840b0ce7d6af2a4daa59/python/global_search_widget/global_search_widget.py#L23